crates/kitsune_p2p/types/src/dht_arc.rs

```diff
@@ -172,147 +389,259 @@ impl RangeBounds<u32> for ArcRange {
+            writeln!(f, "{}", a)?;
+        }
+        writeln!(f, "{} <- Bucket arc", self.filter)
+    }
+}
+
+impl std::fmt::Display for DhtArc {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut out = ["_"; 100];
+        let half_cov = (self.coverage() * 50.0) as isize;
+        let center = self.center_loc.0 .0 as f64 / U32_LEN as f64;
+        let center = (center * 100.0) as isize;
+        for mut i in (center - half_cov)..(center + half_cov) {
+            if i >= 100 {
+                i -= 100;
+            }
+            if i < 0 {
+                i += 100;
             }
-        );
-        check_bounds(0, 2, u32::MAX, 1);
+            out[i as usize] = "#";
+        }
+        out[center as usize] = "|";
+        let out: String = out.iter().map(|a| a.chars()).flatten().collect();
+        writeln!(f, "[{}]", out)
+    }
+}
 
-        assert_eq!(
-            DhtArc::new(u32::MAX, 2).range(),
-            ArcRange {
-                start: Included(u32::MAX - 1),
-                end: Included(0)
+/// Check a set of peers for a gap in coverage.
+/// Note this function is only used for verification in tests at this time.
+pub fn check_for_gaps(peers: Vec<DhtArc>) -> bool {
+    let left = |arc: &DhtArc| match arc.range().start_bound() {
+        Bound::Included(arm) => *arm,
+        _ => unreachable!(),
+    };
+    let mut peers: Vec<_> = peers
+        .into_iter()
+        .filter(|a| matches!(a.range().start_bound(), Bound::Included(_)))
+        .collect();
+    if peers.is_empty() {
+        return true;
+    }
+    peers.sort_unstable_by_key(|p| left(p));
+    if (peers[0].coverage() - 1.0).abs() < ERROR_MARGIN {
+        return false;
+    }
+    // Translate the peers to zero to make wrapping checks easy.
+    let translate = left(&peers[0]);
+    // Safe to cast because of the coverage check
+    let mut max = peers[0].range().len() as u32;
+
+    for peer in peers {
+        if (peer.coverage() - 1.0).abs() < ERROR_MARGIN {
+            return false;
+        }
+        let l = left(&peer) - translate;
+        if l > max {
+            return true;
+        }
+        max = match l.checked_add(peer.range().len() as u32) {
+            Some(m) => m,
+            None => {
+                // We reached the end and we know 0 is covered
+                // so there is no gap.
+                return false;
             }
-        );
-        check_bounds(u32::MAX, 2, u32::MAX - 1, 0);
+        };
+    }
+    // Didn't reach the end
+    true
+}
 
-        assert_eq!(
-            DhtArc::new(0, MAX_HALF_LENGTH).range(),
-            ArcRange {
-                start: Included(half),
-                end: Included(half - 1)
+/// Check a set of peers the actual redundancy across all peers.
+/// This can tell if there is bad distribution.
+/// Note this function is only used for verification in tests at this time.
+pub fn check_redundancy(peers: Vec<DhtArc>) -> usize {
```

Comment:
```suggestion
#[cfg(test)]
pub fn test_check_redundancy(peers: Vec<DhtArc>) -> usize {
```

Same argument as above.